src/main/java/fi/pohina/vinkkilista/App.java

```diff
@@ -106,27 +88,28 @@ public void startServer(int portNumber) {
                 String title = req.queryParams("title");
                 String url = req.queryParams("url");
                 String author = req.queryParams("author");
-                User creator = users.findById(getUserIdFromSession(req.session()));
-                Set<String> tags = tagParser.parse(req.queryParams("tags"));
+                User creator = requestUserManager.getSignedInUser(req);
+                Set<String> tagNames = tagParser.csvToSet(req.queryParams("tags"));
+                tagNames.add(tagService.tagFromUrl(url));
```

Comment:
`App` currently as a controller class coordinates the work of different service classes, `TagParser` and the two `GithubApi` classes.

- This logic responsibility should probably be handled by a separate `AppLogic` class so that `App` has only a single concrete dependency elsewhere
- this class should at least coordinate the three service classes, and perhaps also handle the `GitHub` connection s.t. `App` directly interacts only with `req`, `res` and the logic class
- In case the logic class has too many responsibilities/dependencies, we could have a separate `ServiceCoordinator` for the service classes as well as a `Github` class for `Json` and `OAuth` `GithubApi`
- `TagParser` could alternatively belong to `TagService`

Furthermore, I think delegating the get/post functionality to helper methods would be an improvement, too many things things are happening in get "/new" right now (worst offender).

Reply:
I'm not sure about moving everything to a single `AppLogic` service, since that would just be a thin wrapper around the other services...which may just as well be injected individually. The Github API interactions could definitely be moved to a `GithubService` since right now the API classes do a lot more than just interact with the APIs. Some of the responsibilities found there currently should probably be something that a service layer class does.

Reply:
But yeah, TagParser's things belong to the service IMO, I agree.